tasks/scheduler.go

```diff
@@ -68,43 +49,75 @@ func NewScheduler(workers int, deliverTxFunc func(ctx sdk.Context, req types.Req
 	}
 }
 
-func (s *scheduler) ProcessAll(ctx sdk.Context, tasks []*Task) error {
+func toTasks(reqs []types.RequestDeliverTx) []*deliverTxTask {
+	res := make([]*deliverTxTask, 0, len(reqs))
+	for idx, r := range reqs {
+		res = append(res, &deliverTxTask{
+			Request: r,
+			Index:   idx,
+			Status:  statusPending,
+		})
+	}
+	return res
+}
+
+func collectResponses(tasks []*deliverTxTask) []types.ResponseDeliverTx {
+	res := make([]types.ResponseDeliverTx, 0, len(tasks))
+	for _, t := range tasks {
+		res = append(res, *t.Response)
+	}
+	return res
+}
+
+func (s *scheduler) ProcessAll(ctx sdk.Context, reqs []types.RequestDeliverTx) ([]types.ResponseDeliverTx, error) {
+	tasks := toTasks(reqs)
 	toExecute := tasks
 	for len(toExecute) > 0 {
 
 		// execute sets statuses of tasks to either executed or aborted
 		err := s.executeAll(ctx, toExecute)
 		if err != nil {
-			return err
+			return nil, err
 		}
 
 		// validate returns any that should be re-executed
 		// note this processes ALL tasks, not just those recently executed
 		toExecute, err = s.validateAll(ctx, tasks)
 		if err != nil {
-			return err
+			return nil, err
 		}
 		for _, t := range toExecute {
 			t.Incarnation++
-			t.Status = TaskStatusPending
+			t.Status = statusPending
+			t.Response = nil
 			//TODO: reset anything that needs resetting
 		}
 	}
-	return nil
+	return collectResponses(tasks), nil
 }
 
-// TODO: validate each task
+// TODO: validate each tasks
 // TODO: return list of tasks that are invalid
-func (s *scheduler) validateAll(ctx sdk.Context, tasks []*Task) ([]*Task, error) {
-	var res []*Task
-	for _, t := range tasks {
+func (s *scheduler) validateAll(ctx sdk.Context, tasks []*deliverTxTask) ([]*deliverTxTask, error) {
```

Comment:
this isnt actually doing any validation, should we rename this to something like collectTasksForValidation (since thats what it appears to be doing?)

Reply:
it's going to be doing the validation, it's collecting tasks for re-execution